_posts/2022-11-29-November-Update-Exploring-V3D.md

```diff
@@ -0,0 +1,67 @@
+
+Next, there are Performance Monitor (perfmon) related-ioctls: `v3d_perfmon_create`, `v3d_perfmon_destroy`, and `v3d_perfmon_get_values`. Performance Monitors are basically registers that are used for monitoring the performance of the V3D system. So, tests were designed to assure that the driver was creating perfmons properly and was resilient to incorrect requests, such as trying to get a value from a non-existent perfmon.
+
+And finally, the most interesting type of ioctls: the job submission ioctls. You can use the `v3d_submit_cl` ioctl to submit commands to the 3D engine, which is a tiled engine. When I think about tiled rendering, I always think about a [Super Nintendo](https://www.youtube.com/@RGMechEx), but things can get a bit more complicated than a SNES as you can see [here](https://developer.samsung.com/galaxy-gamedev/resources/articles/gpu-framebuffer.html). The 3D engine is composed of a bin and render pipelines, each has its command list and the render depends on the binner. By testing the v3d_submit_cl ioctl, it is possible to test syncing between jobs and also the [V3D multisync ability](https://melissawen.github.io/blog/2022/05/10/multisync-p1).
+
+Moreover, the V3D has also a TFU (texture formatting unit), and a CSD (compute shader dispatch), which has their ioctls: `v3d_submit_tfu` and `v3d_submit_csd`. The TFU makes format conversions and generated mipmaps and the CSD is responsible for dispatching a compute shader.
+
+So, the idea is to write tests for all those functionalities from V3D, increasing the testability of V3D on IGT.
+
+## A Noop Job...
+
+In order to test the `v3d_submit_cl` ioctl, it was needed to design a job to be submitted. So, Melissa suggested using Mesa's noop job specification on IGT to perform the tests. The idea was quite simple: submit a noop job and create tests based on it. But, it was not that simple after all...
+
+First, I must say that I'm mostly a kernel developer, so I was not familiar with Mesa. So, maybe it was not that hard to figure out, but I took a while to understand Mesa's packet and how to submit them.
+
+The main problem I faced on submitting a noop job on IGT was: I would copy many and many Mesa files to IGT. And I took a while fighting against this idea, looking for other ways to submit a job to V3D. But, as all experience developers pointed out, packeting is the best option for it.
+
+After some time, I was able to bring the Mesa structure to IGT with a minimal (although not that minimal) overhead. But, I'm still not able to run a successful noop job as the job's fence is not being signaled by the end of the job.
+
+Although things are not yet working, it was pretty nice to explore Mesa and learn more from it. After this experience with Mesa and also XDC, I became more and more interested in Mesa.
+
+## Series Submitted
```

Comment:
I think this blog post should focus on the work you've already done. In that sense, I would move this section "series submitted" to a place before the line 37, wrapping up your work on the first deliverables (all those tests and improvements). Then I see the `v3d_submit_cl` is part of your "next steps" and you can explain your on-going development and what you plan to deliver. What do you think?